inductor: add params check before doing sfdp fusion

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #100619
* #100612

We need to add params check before doing sfdp fusion as https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/attention.cpp#L544.

This PR will fix #100315 and #100318.


cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
inductor: add params check before doing sfdp fusion

[ghstack-poisoned]

XiaobingSuper committed:
Update on "inductor: add params check before doing sfdp fusion"


We need to add params check before doing sfdp fusion as https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/attention.cpp#L544.

This PR will fix #100315.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -165,7 +165,7 @@ def _sfdp_params_check(match):
         if (
             not isinstance(attn_mask, torch.Tensor)
-            or (attn_mask.dtype != query.dtype and attn_mask.dtype != torch.bool)
+            or attn_mask.dtype != query.dtype
```
Update the comment at https://github.com/pytorch/pytorch/pull/100619/files/a177cfadac8b75e97dc7bd412f9509adb0c30f88..8cf3142ec663db6109796e6a22ead27c5f6283d8#diff-d2f5fd73265c51de38cecb3b0309a2e555e8260feb7bd16a963fd9b89321a537L162 too?

XiaobingSuper commented on torch/_inductor/fx_passes/fuse_attention.py:
```diff
@@ -146,6 +146,32 @@ def _sfdp_replacement_6(query, key, value, attn_mask, dropout_p):
+        if (
+            not isinstance(attn_mask, torch.Tensor)
+            or attn_mask.dtype != query.dtype
+            or query.device != attn_mask.device
```
Currently, we need to make sure ```attn_mask.dtype == query.dtype```, if we meet ```attn_mask.dtype != query.dtype``` in a real model, we can add dtype conversion before calling ```scaled_dot_product_attention```.

XiaobingSuper committed:
Update on "inductor: add params check before doing sfdp fusion"


We need to add params check before doing sfdp fusion as https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/transformers/attention.cpp#L544.

This PR will fix #100315 and #100318.


cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
